### PR TITLE
[Pornbox] New extractor

### PR DESCRIPTION
**Pull request by niemands**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This extractor adds support for videos on pornbox.com. Most of the videos on the site must be purchased individually or are accessible via channel membership options. The test case is for one of the few free promotional videos. However, it still requires an account.
This extractor can return metadata with `--dump-json` and `--ignore-no-formats`, but otherwise requires cookie authentication.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🚀🗣️🛠️

Add support for `--write-link` and `--write-url` options to write video or playlist URLs to a file. Refactor the code for writing metadata files to avoid duplication and improve readability.

> _`get` and `set` methods_
> _refactored for readability_
> _autumn leaves falling_

### Walkthrough
*  Add new extractors for various video platforms (F0L

</details>

### CODE REVIEW

File: yt_dlp/extractor/pornbox.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            'cast': ['Bella Grey', 'Oliver Trunk'],
+            'tags': 'count:29',
+            'thumbnail': r're:^https?://cdn-image\.gtflixtv\.com.*\.jpg.*$',
+            'subtitles': 'count:6'
+        },
+        'params': {
+            'skip_download': True,
+            'ignore_no_formats_error': True
+        },
+        'skip': 'Only for subscribers'
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:55]: You don't need to skip if doing ignore_no_formats_error, no?

File: yt_dlp/extractor/pornbox.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            cast.append(m.get('model_name'))
+        for m in public_data.get('male_models') or []:
+            cast.append(m.get('model_name'))
+
+        subtitles = {}
+        for country_code in public_data.get('subtitles') or []:
+            subtitles[country_code] = [{
+                'url': f'https://pornbox.com/contents/{video_id}/subtitles/{country_code}',
+                'ext': 'srt'
+            }]
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:75]:
```suggestion
        subtitles = {country_code: [{
            'url': f'https://pornbox.com/contents/{video_id}/subtitles/{country_code}',
            'ext': 'srt'
        }] for country_code in traverse_obj(public_data, ('subtitles', ...)}
```

File: yt_dlp/extractor/pornbox.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                                      metadata_available=True, method='cookies')
+            return metadata
+
+        media_id = traverse_obj(public_data, ('medias', lambda _, v: v['title'] == 'Full video', 'media_id'),
+                                default=[], expected_type=int)
+        if not media_id:
+            self.raise_no_formats(msg='Could not find stream id', video_id=video_id)
+
+        # traverse_obj branches, therefore media_id is a list
+        media_id = media_id[0]
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:108]:
```suggestion
        media_id = traverse_obj(
            public_data, ('medias', lambda _, v: v['title'] == 'Full video', 'media_id', {int}), get_all=False)
        if not media_id:
            self.raise_no_formats('Could not find stream id', video_id=video_id)
```

File: yt_dlp/extractor/pornbox.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            'Accept': '*/*',
+            'Accept-Language': 'en-US;q=0.7,en;q=0.3',
+            'Accept-Encoding': 'gzip, deflate, br',
+            'X-Requested-With': 'XMLHttpRequest',
+            'DNT': '1',
+            'Connection': 'keep-alive',
+            'Sec-Fetch-Dest': 'empty',
+            'Sec-Fetch-Mode': 'cors',
+            'Sec-Fetch-Site': 'same-origin'
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:120]: Most of these are unnecessary

File: yt_dlp/extractor/pornbox.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        stream_data = self._download_json(f'https://pornbox.com/media/{media_id}/stream', video_id=video_id,
+                                          headers=headers, note='Getting manifest urls')
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:122]:
```py
        stream_data = self._download_json(
            f'https://pornbox.com/media/{media_id}/stream', video_id, 'Getting manifest urls',
            headers={
                ....
            })
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/pornbox.py (hunk 1 above)
Comment (niemands) [reply yt_dlp/extractor/pornbox.py:55]: I tried, but the test aborts with the login (raise_login_required) error

File: yt_dlp/extractor/pornbox.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        video_id = self._match_id(url)
+
+        public_data = self._download_json(f'https://pornbox.com/contents/{video_id}', video_id)
+
+        date = traverse_obj(public_data, ('studios', 'release_date'), 'publish_date')
+        cast = []
+        for m in public_data.get('models') or []:
+            cast.append(m.get('model_name'))
+        for m in public_data.get('male_models') or []:
+            cast.append(m.get('model_name'))
```
[the remark this answers is not in the document]
Comment (niemands) [reply yt_dlp/extractor/pornbox.py:68]: Wouldn't this return the full `public_data` object?

The public_data json structure is 
```json
{
  ...
  "models": [
    {
      "model_name": "name1",
      ...
    }
  ],
  "male_models": [
    {
      "model_name": "name2",
      ...
    }
  ],
  ...
}
```

I have tried to create an inline expression like
`cast  = traverse_obj(public_data, ('models', ..., 'model_name'), ('male_models', ..., 'model_name'))` but the results don't get merged.
The closest I get to a single line expression would be this one, I don't know if this is better though:
```suggestion
        cast = traverse_obj(public_data, ('models', ..., 'model_name')) + traverse_obj(public_data, ('male_models', ..., 'model_name'))
```

[the remark this answers is not in the document]
Comment (bashonly) [reply yt_dlp/extractor/pornbox.py:68]:
```py
        cast = traverse_obj(public_data, (('models', 'male_models'), ..., 'model_name'))
```

File: yt_dlp/extractor/pornbox.py (hunk 4 above)
Comment (niemands) [reply yt_dlp/extractor/pornbox.py]: You are right, removed them all

File: yt_dlp/extractor/pornbox.py (hunk 1 above)
Comment (niemands) [reply yt_dlp/extractor/pornbox.py:54]: I think I have found the problem. `raise_login_required()` calls `self.report_warning()`, which is modified in line 51 of test_download.py to throw an `ExtractorError`. This is then caught in line 160. Based on the comment in line 50 ("Don't accept warnings during tests"), I would assume that this is intended behaviour and nothing I can work around here.

I have temporarily commented out the part where the `ExtractorError` is thrown and then the test passes.

Comment (bashonly) [reply yt_dlp/extractor/pornbox.py:54]: I think it needs this
```suggestion
        'expected_warnings': ['This video is only available for registered users'],
        'skip': 'Only for subscribers'
```

File: yt_dlp/extractor/pornbox.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/pornbox.py]: Mb. Use bashonly's

File: yt_dlp/extractor/pornbox.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        video_id = self._match_id(url)
+
+        public_data = self._download_json(f'https://pornbox.com/contents/{video_id}', video_id)
+
+        subtitles = {country_code: [{
+            'url': f'https://pornbox.com/contents/{video_id}/subtitles/{country_code}',
+            'ext': 'srt'
+        }] for country_code in traverse_obj(public_data, ('subtitles', ...))}
+
+        is_free_scene = traverse_obj(public_data, ('price', 'is_available_for_free'), default=False, expected_type=bool)
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:67]:
```suggestion
        is_free_scene = traverse_obj(
            public_data, ('price', 'is_available_for_free', {bool}), default=False)
```

File: yt_dlp/extractor/pornbox.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            'age_limit': 18,
+            'duration': parse_duration(public_data.get('runtime')),
+            'cast': traverse_obj(public_data, (('models', 'male_models'), ..., 'model_name')),
+            'tags': traverse_obj(public_data, ('niches', ..., 'niche'), default=[]),
+            'thumbnail': str_or_none(public_data.get('player_poster')),
+            'subtitles': subtitles,
+            'availability': self._availability(needs_auth=True, needs_premium=(not is_free_scene))
+        }
+
+        if (not public_data.get('is_purchased')) or (not is_free_scene):
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:84]:
```suggestion
        if not public_data.get('is_purchased') or not is_free_scene:
```

**Review by pukkandan — APPROVED** [review]
nitpicks

File: yt_dlp/extractor/pornbox.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            'uploader': public_data.get('studio'),
+            'timestamp': parse_iso8601(traverse_obj(public_data, ('studios', 'release_date'), 'publish_date')),
+            'age_limit': 18,
+            'duration': parse_duration(public_data.get('runtime')),
+            'cast': traverse_obj(public_data, (('models', 'male_models'), ..., 'model_name')),
+            'tags': traverse_obj(public_data, ('niches', ..., 'niche'), default=[]),
+            'thumbnail': str_or_none(public_data.get('player_poster')),
+            'subtitles': subtitles,
+            'availability': self._availability(needs_auth=True, needs_premium=(not is_free_scene))
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:85]:
```suggestion
        metadata = {
            'id': video_id,
            **traverse_obj(public_data, {
                'title': ('scene_name', {str.strip}),
                'description': ('small_description', {str.strip}),
                'uploader': 'studio',
                'duration': ('runtime', {parse_duration}),
                'cast': (('models', 'male_models'), ..., 'model_name'),
                'thumbnail': ('player_poster', {url_or_none}),
                'tags': ('niches', ..., 'niche'),
            }),
            'age_limit': 18,
            'timestamp': parse_iso8601(traverse_obj(
                public_data, ('studios', 'release_date'), 'publish_date'))
            'availability': self._availability(needs_auth=True, needs_premium=(not is_free_scene))
            'subtitles': subtitles,
        }
```

File: yt_dlp/extractor/pornbox.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
+from .common import InfoExtractor
+from ..compat import functools
+from ..utils import (
+    parse_duration,
+    traverse_obj,
+    str_or_none,
+    int_or_none,
+    strip_or_none,
+    parse_iso8601,
+    qualities
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:10]:
```suggestion
    parse_iso8601,
    qualities,
    url_or_none,
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/pornbox.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                'uploader': 'studio',
+                'duration': ('runtime', {parse_duration}),
+                'cast': (('models', 'male_models'), ..., 'model_name'),
+                'thumbnail': ('player_poster', {url_or_none}),
+                'tags': ('niches', ..., 'niche'),
+            }),
+            'age_limit': 18,
+            'timestamp': parse_iso8601(traverse_obj(
+                public_data, ('studios', 'release_date'), 'publish_date'))
+            'availability': self._availability(needs_auth=True, needs_premium=(not is_free_scene))
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:86]:
```suggestion
            'availability': self._availability(needs_auth=True, needs_premium=(not is_free_scene)),
```

File: yt_dlp/extractor/pornbox.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+                'description': ('small_description', {str.strip}),
+                'uploader': 'studio',
+                'duration': ('runtime', {parse_duration}),
+                'cast': (('models', 'male_models'), ..., 'model_name'),
+                'thumbnail': ('player_poster', {url_or_none}),
+                'tags': ('niches', ..., 'niche'),
+            }),
+            'age_limit': 18,
+            'timestamp': parse_iso8601(traverse_obj(
+                public_data, ('studios', 'release_date'), 'publish_date'))
```
Comment (pukkandan) [inline yt_dlp/extractor/pornbox.py:85]:
```suggestion
            'timestamp': parse_iso8601(traverse_obj(
                public_data, ('studios', 'release_date'), 'publish_date')),
```

File: yt_dlp/extractor/pornbox.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
+from .common import InfoExtractor
+from ..compat import functools
+from ..utils import (
+    parse_duration,
+    traverse_obj,
+    str_or_none,
+    int_or_none,
+    parse_iso8601,
+    qualities,
+    url_or_none,
```
Comment (bashonly) [inline yt_dlp/extractor/pornbox.py:10]:
```suggestion
    int_or_none,
    parse_duration,
    parse_iso8601,
    qualities,
    str_or_none,
    traverse_obj,
    url_or_none,
```

File: yt_dlp/extractor/pornbox.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        public_data = self._download_json(f'https://pornbox.com/contents/{video_id}', video_id)
+
+        subtitles = {country_code: [{
+            'url': f'https://pornbox.com/contents/{video_id}/subtitles/{country_code}',
+            'ext': 'srt'
+        }] for country_code in traverse_obj(public_data, ('subtitles', ...))}
```
Comment (bashonly) [inline yt_dlp/extractor/pornbox.py:67]:
```suggestion
        subtitles = {country_code: [{
            'url': f'https://pornbox.com/contents/{video_id}/subtitles/{country_code}',
            'ext': 'srt'
        }] for country_code in traverse_obj(public_data, ('subtitles', ..., {str}))}
```

File: yt_dlp/extractor/pornbox.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+            'age_limit': 18,
+            'timestamp': parse_iso8601(traverse_obj(
+                public_data, ('studios', 'release_date'), 'publish_date')),
+            'availability': self._availability(needs_auth=True, needs_premium=not is_free_scene),
+            'subtitles': subtitles,
+        }
+
+        if not public_data.get('is_purchased') or not is_free_scene:
+            self.raise_login_required('You are either not logged in or do not have access to this scene',
+                                      metadata_available=True, method='cookies')
```
Comment (bashonly) [inline yt_dlp/extractor/pornbox.py:92]:
```suggestion
            self.raise_login_required(
                'You are either not logged in or do not have access to this scene', metadata_available=True)
```

File: yt_dlp/extractor/pornbox.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+            f'https://pornbox.com/media/{media_id}/stream', video_id=video_id, note='Getting manifest urls')
+
+        get_quality = qualities(['web', 'vga', 'hd', '1080p', '4k', '8k'])
+        metadata['formats'] = [traverse_obj(q, {
+            'url': 'src',
+            'vbr': ('bitrate', {functools.partial(int_or_none, scale=1000)}),
+            'format_id': ('quality', {str_or_none}),
+            'quality': ('quality', {get_quality}),
+            'width': ('size', {lambda x: int(x[:-1])})
+        }) for q in stream_data.get('qualities') or []]
```
Comment (bashonly) [inline yt_dlp/extractor/pornbox.py:110]:
```suggestion
        metadata['formats'] = traverse_obj(stream_data, ('qualities', lambda _, v: v['src'], {
            'url': 'src',
            'vbr': ('bitrate', {functools.partial(int_or_none, scale=1000)}),
            'format_id': ('quality', {str_or_none}),
            'quality': ('quality', {get_quality}),
            'width': ('size', {lambda x: int(x[:-1])}),
        }))
```

**Review by bashonly — APPROVED** [review]

**Comment by niemands** [comment]
I had to replace the first test case with a new video because the old one was no longer available, and as far as I can see this is one of the four remaining free videos, so I'm not sure if this test case will work for long. However, it passes the test.

**Pull request merged**